hypothesis-python/src/hypothesis/internal/floats.py

```diff
@@ -99,3 +100,22 @@ def next_up(value, width=64):
 
 def next_down(value, width=64):
     return -next_up(-value, width)
+
+
+# Smallest positive non-zero numbers that is fully representable by an
+# IEEE-754 float, calculated with the width's associated minimum exponent.
+# Values from https://en.wikipedia.org/wiki/IEEE_754#Basic_and_interchange_formats
+width_smallest_normals = {
+    16: 2 ** -(2 ** (5 - 1) - 2),
+    32: 2 ** -(2 ** (8 - 1) - 2),
+    64: 2 ** -(2 ** (11 - 1) - 2),
+}
+assert width_smallest_normals[64] == float_info.min
+
+
+# Specifies whether we can represent subnormal floating point numbers.
+# IEE-754 requires subnormal support, but it's often disabled anyway by unsafe
+# compiler options like `-ffast-math`.  On most hardware that's even a global
+# config option, so *linking against* something built this way can break us.
+# Everything is terrible
+PYTHON_FTZ = next_down(float_info.min) == 0.0
```

Comment:
I am inclined to just `assert next_down(float_info.min) > 0.0, "flush-to-zero violates IEEE-754"` here and hope we never have to deal with it.

Reply:
Haha, yeah I see that.

The scenarios where folk are using FTZ Python is seemingly very rare. That said it'd be obviously nice to support it if possible, and maybe there's a scenario in the future where some distributor of Python bins/packages starts building with say `-ffast-math`*. Hell if a library like CuPy distributes to FTZ by default for performance gains, maybe a future Python flavour will violate IEEE-754 and build FTZ by default.

So uhh, a lot of ifs. Could remove it for now and just crawl through the history if FTZ Python becomes a real use-case. If left as-is, I could at least add a warning.

Note `-ffast-math` seems inconsistent with FTZ behaviour depending on the compiler and other flags, why I was having problems reproducing FTZ NumPy.

Reply:
I just think this scenario is rare enough that it's not worth adding code to handle it, especially in the context of an already-huge PR.  Forget the assertion, let's just stick with `test_compiled_with_sane_math_options()` in our tests and otherwise ignore the possibility.

And yeah, the compiler flags are a bit of a mess.  Sorry about that 😞

Reply:
> not worth adding code to handle it, especially in the context of an already-huge PR

- [ ] So I'll remove the `allow_subnormal=False` defaulting and warning in `st.floats()` that uses `PYTHON_FTZ`.

> let's just stick with `test_compiled_with_sane_math_options`

- [ ] Reinstate this test (I had removed it heh)